src/pybind/rados/rados.pyx

```diff
@@ -2756,6 +2820,44 @@ returned %d, but should return zero on success." % (self.name, ret))
             raise LogicError("Ioctx.write_full(%s): rados_write_full \
 returned %d, but should return zero on success." % (self.name, ret))
 
+    @requires(('key', str_type), ('data', bytes), ('write_len', int), ('offset', int))
+    def writesame(self, key, data, write_len, offset=0):
+        """
+        Write the same buffer multiple times
+        :param key: name of the object
+        :type key: str
+        :param data: data to write
+        :type data: bytes
+        :param len: total number of bytes to write
+        :type len: int
+        :param offset: byte offset in the object to begin writing at
+        :type offset: int
+
+        :raises: :class:`TypeError`
+        :raises: :class:`LogicError`
+        :returns: int - 0 on success
```

Comment:
don't think we'd need something to be returned here, to keep things simple.

Reply:
```
int rados_writesame(rados_ioctx_t io, const char *oid, const char *buf, size_t data_len, size_t write_len, uint64_t off)
: returns 0 on success, negative error code on failure
```
we need to catch exceptions.

Reply:
we are not using the return `ret` here for any purpose that would be confusing to have, would be more relevant to make this function return nothing, and for raising an error you can do something like : 
```
if ret != 0:
    raise make_ex(ret, "Ioctx.writesame(%s): failed to write %s"
                           % (self.name, key))
```

Reply:
"ret > 0"  and "ret < 0" have different meanings.

Reply:
I think if the code fails to execute it will throw an error, so do we need to return a `0` in this function if you can explain or change that will help. Other than that LGTM.

Reply:
```
       with nogil:
            ret = rados_writesame(self.io, _key, _data, _data_len, _write_len, _offset)
        if ret == 0:
            return ret
        elif ret < 0: 
            raise make_ex(ret, "Ioctx.writesame(%s): failed to write %s"
                           % (self.name, key))
        else:
            raise LogicError("Ioctx.(%s): rados_writesame \
returned %d, but should return zero on success." % (self.name, ret))
```
@ideepika   If there is any mistake, please point it out and I will revise it. If not, please approve it. Thanks a lot.

Reply:
@zjcmszh i concur with Deepika. 

per the document of `rados_writesame()`. there are two cases:

>  * @returns 0 on success, negative error code on failure

we could handle the positive number case by returning an `assert()` or by throwing an exception. both is fine. as if the callee breaks the contract, the behavior is undefined.

in the case of a negative return number, an customized exception is thrown in your change. that's good. as it's the way how python returns an unexpected error.

but when it comes the 0 return value, you make it a special case by returning "0". so what the caller of `self.ioctx.writesame()` is supposed to check the return code? by comparing the return code with "0"? that's silly, i'd say. as the only expected return value is "0". and i think the reason why your test does not check the return code at all is also based on this assumption.

so, i'd suggest put something like
```py
if ret < 0:
  raise make_ex(ret, "Ioctx.writesame(%s): failed to write %s" % (self.name, key))
assert(ret == 0)
```